dns_sd_windows.c

```diff
@@ -140,52 +216,140 @@ static int query_callback(int sock, const struct sockaddr *from, size_t addrlen,
+		goto quit;
 
-	IIO_DEBUG("DNS SD: added %s (%s:%d)\n",
-		  dd->hostname, dd->addr_str, dd->port);
+	mdns_string_t fromaddrstr =
+			ip_address_to_string(addrbuffer, sizeof(addrbuffer), from, addrlen);
+
+	iio_mutex_lock(dd->lock);
+	if (rtype == MDNS_RECORDTYPE_SRV) {
+		srv = mdns_record_parse_srv(data, size, record_offset, record_length,
+			namebuffer, sizeof(namebuffer));
+		IIO_DEBUG("%.*s : %.*s SRV %.*s priority %d weight %d port %d\n",
+			MDNS_STRING_FORMAT(fromaddrstr), MDNS_STRING_FORMAT(entrystr),
+			MDNS_STRING_FORMAT(srv.name), srv.priority, srv.weight, srv.port);
+
+		/* find a match based on name/port/ipv4 or 6 & update it, otherwise add it */
+		while (dd->next) {
+			if (dd->hostname && !strncmp(dd->hostname, srv.name.str, srv.name.length - 1) &&
+					(!dd->port || dd->port == srv.port) && dd->found == (from->sa_family != AF_INET)) {
+				dd->port = srv.port;
+				IIO_DEBUG("DNS SD: updated SRV %s (%s port:%hu)\n", dd->hostname, dd->addr_str, dd->port);
+				found = true;
+			}
+			dd = dd->next;
+		}
+		if (!found) {
+			/* new hostname and port */
+			if (srv.name.length > 1) {
+				dd->hostname = iio_strndup(srv.name.str, srv.name.length - 1);
+				if (!dd->hostname)
+					goto mem_fail;
+			}
 
-	/* A list entry was filled, prepare new item on the list */
-	dd->next = zalloc(sizeof(*dd->next));
-	if (!dd->next)
-		IIO_ERROR("DNS SD mDNS Resolver : memory failure\n");
+			iio_strlcpy(dd->addr_str, fromaddrstr.str, fromaddrstr.length + 1);
+			dd->port = srv.port;
+			dd->found = (from->sa_family != AF_INET);
+			IIO_DEBUG("DNS SD: added SRV %s (%s port:%d)\n",
+				dd->hostname, dd->addr_str, dd->port);
 
+			/* A list entry was filled, prepare new item on the list */
+			dd->next = new_discovery_data(dd);
+			if (!dd->next)
+				goto mem_fail;
+		}
+	} else if (rtype == MDNS_RECORDTYPE_A) {
+		struct sockaddr_in addr;
+		mdns_record_parse_a(data, size, record_offset, record_length, &addr);
+		mdns_string_t addrstr =
+			ipv4_address_to_string(namebuffer, sizeof(namebuffer), &addr, sizeof(addr));
+		IIO_DEBUG("%.*s : %.*s A %.*s\n", MDNS_STRING_FORMAT(fromaddrstr),
+			MDNS_STRING_FORMAT(entrystr), MDNS_STRING_FORMAT(addrstr));
+
+		/* find a match based on name/ipv4 or 6 & update it, otherwise add it */
+		while (dd->next) {
+			if (dd->hostname && !strncmp(dd->hostname, entrystr.str, entrystr.length - 1) &&
+					!dd->found) {
+				iio_strlcpy(dd->addr_str, addrstr.str, addrstr.length + 1);
+				IIO_DEBUG("DNS SD: updated A %s (%s port:%hu)\n", dd->hostname, dd->addr_str, dd->port);
+				found = true;
+			}
+			dd = dd->next;
+		}
+		if (!found) {
+			dd->hostname = iio_strndup(entrystr.str, entrystr.length - 1);
+			if (!dd->hostname)
+				goto mem_fail;
+			iio_strlcpy(dd->addr_str, addrstr.str, addrstr.length + 1);
+			dd->found = 0;
+
+			/* A list entry was filled, prepare new item on the list */
+			dd->next = new_discovery_data(dd);
+			if (!dd->next)
+				goto mem_fail;
+		}
+	} 
```

Comment:
There's an extra space after the closing bracket here.

Reply:
fixed